File: services/api-ts/src/adapters/memory-task-repository.ts
import type { TaskRepository } from "../application/ports.ts";
import type { Task } from "../domain/task.ts";

/**
 * In-process TaskRepository: the store the service starts with, and the behaviour any other store
 * must match. A Map keeps insertion order, and replacing a key keeps its original position.
 */
export class MemoryTaskRepository implements TaskRepository {
  readonly #tasks = new Map<string, Task>();

  async save(task: Task): Promise<void> {
    this.#tasks.set(task.id, task);
  }

  async get(id: string): Promise<Task | undefined> {
    return this.#tasks.get(id);
  }

  async list(): Promise<readonly Task[]> {
    return [...this.#tasks.values()];
  }
}

File: services/api-py/src/api_py/adapters/memory_task_repository.py
"""The in-memory store: what the service starts with, and the reference behaviour any other store
must match. It forgets everything on restart.

A dict preserves insertion order, so ``list`` returns tasks oldest first without keeping a second
index; a re-saved task keeps its original position, as it does in api-go and api-ts.
"""

from __future__ import annotations

import threading

from api_py.domain.task import Task


class MemoryTaskRepository:
    """Satisfies the TaskRepository port structurally; tests assert the port is satisfied."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._tasks: dict[str, Task] = {}

    def save(self, task: Task) -> None:
        with self._lock:
            self._tasks[task.id] = task

    def get(self, task_id: str) -> Task | None:
        with self._lock:
            return self._tasks.get(task_id)

    def list(self) -> tuple[Task, ...]:
        with self._lock:
            return tuple(self._tasks.values())
